Only trigger changes for tables, cards and calendars if necessary (otherwise triggers auto-save)

leon-marzahn opened the pull request:
## Scope

What's changed:

- The query limit in a table/cards view is only updated if the QUERY_LIMIT_MAX would actually change the query limit on said table
- In a calendar view, only subscribe to `datesSet` after rendering, skipping the initial value (as the initial value is the same before and after rendering)
- ~~Don't create/save preset on page change~~ Reverted by https://github.com/directus/directus/pull/20106/commits/bea5cf26122e9f9940588c0da21a1d64e08ee64d, as per discussion in https://github.com/directus/directus/pull/20106/files#r1374908744
- Fixes unintended behaviour where a preset is saved even when no values were changed

## Potential Risks / Drawbacks

- There shouldn't be any for table/cards views, but welcome to a sanity check
- I'm unsure about the calendar behaviour and the swap of render and subscription might cause unintended effects

## Questions

- I wasn't able to investigate whether kanban and map views have similar behaviour, so might be worth looking into in the future

---

Fixes #20105

leon-marzahn committed:
Only change items per page in table if necessary (otherwise triggers auto-save)

leon-marzahn committed:
Only change items per page in cards if necessary (otherwise triggers auto-save)

leon-marzahn committed:
Postpone view info update in calendar (otherwise triggers auto-save)

leon-marzahn committed:
Sign CLA

leon-marzahn commented:
JFYI i have no prior experience with vue, so please let me know if i did something wrong :)

leon-marzahn committed:
Fix formatting

leon-marzahn committed:
Merge branch 'main' into no-preset-auto-save

leon-marzahn committed:
Merge branch 'main' into no-preset-auto-save

leon-marzahn committed:
Update contributors.yml

leon-marzahn committed:
Merge branch 'main' into no-preset-auto-save

leon-marzahn committed:
Merge branch 'main' into no-preset-auto-save

paescuj committed:
Combine changesets

paescuj reviewed (dismissed):
Great findings @leon-marzahn ❤️ LGTM 🚀

leon-marzahn commented:
@paescuj Thanks ❤️

br41nslug reviewed (changes requested):
At least for the table layout this does not seem to be a complete fix for the original issue.

https://github.com/directus/directus/assets/9389634/d249bc21-f43b-4322-ac4f-9f90ec6549a0

leon-marzahn commented:
> At least for the table layout this does not seem to be a complete fix for the original issue.
> 
>  firefox_bz2RUikqto.mp4

It might be related to the fact there weren't any presets beforehand, will investigate now!

paescuj commented:
> At least for the table layout this does not seem to be a complete fix for the original issue.
> firefox_bz2RUikqto.mp4

Interesting catch! This happens when you switch between collections and the `page` in `layoutQuery` changes.
We could (and maybe should) filter that out, but that feels out of scope for this issue / PR since the original concern was about unnecessary creation of additional presets even though a preset already existed.

br41nslug commented:
Interesting will do some further testing then 😄

paescuj commented:
> Interesting will do some further testing then 😄

Cool ❤️ Here's the corresponding line:
https://github.com/directus/directus/blob/c80f8ee2476f847b3df9d5238ca2f5a37496e457/app/src/composables/use-preset.ts#L121

I wouldn't consider this as a blocker for this PR though & prefer to track this in a separate issue.

leon-marzahn commented:
> > Interesting will do some further testing then 😄
> 
> Cool ❤️ Here's the corresponding line:
> 
> https://github.com/directus/directus/blob/c80f8ee2476f847b3df9d5238ca2f5a37496e457/app/src/composables/use-preset.ts#L121
> 
> I wouldn't consider this as a blocker for this PR though & prefer to track this in a separate issue.

It seems to be these lines in particular. They compare the difference between the previous table and current table and set the page to 1 if they are different in any way. I wonder if it makes more sense to just set the page to 1 if it isn't 1 already?

But even then, if you were on a different page it would save a preset even though you have just moved to a different page, that seems unintended?

https://github.com/directus/directus/blob/c80f8ee2476f847b3df9d5238ca2f5a37496e457/packages/composables/src/use-items.ts#L89-L98

leon-marzahn commented:
> > > Interesting will do some further testing then 😄
> > 
> > 
> > Cool ❤️ Here's the corresponding line:
> > https://github.com/directus/directus/blob/c80f8ee2476f847b3df9d5238ca2f5a37496e457/app/src/composables/use-preset.ts#L121
> > 
> > I wouldn't consider this as a blocker for this PR though & prefer to track this in a separate issue.
> 
> It seems to be these lines in particular. They compare the difference between the previous table and current table and set the page to 1 if they are different in any way. I wonder if it makes more sense to just set the page to 1 if it isn't 1 already?
> 
> But even then, if you were on a different page it would save a preset even though you have just moved to a different page, that seems unintended?
> 
> https://github.com/directus/directus/blob/c80f8ee2476f847b3df9d5238ca2f5a37496e457/packages/composables/src/use-items.ts#L89-L98

I guess i've looked to deep than i needed to, the easiest solution would be ignoring page changes 😅 I will create another issue for that, unless it's easier to just include here as it does sort-of fit.

leon-marzahn commented:
@paescuj @br41nslug This is a very ugly solution, but something like would work:
```
function updatePreset(preset: Partial<Preset>, immediate?: boolean) {
    localPreset.value = assign({}, localPreset.value, preset);

    let hasChanged = false;

    for (const key in preset.layout_query ?? {}) {
	if (Object.keys((preset.layout_query ?? {})[key]).length === 1 && typeof (preset.layout_query ?? {})[key]?.page !== 'undefined') {
		continue;
	}

	hasChanged = true;
    }

    if (!hasChanged) return;
    immediate ? savePreset() : handleChanges();
}
```

paescuj commented:
> [...] I wonder if it makes more sense to just set the page to 1 if it isn't 1 already?
> But even then, if you were on a different page it would save a preset even though you have just moved to a different page, that seems unintended?

This is why I prefer to create a separate issue for this, so we can discuss such things there 😄

leon-marzahn commented:
> > [...] I wonder if it makes more sense to just set the page to 1 if it isn't 1 already?
> > But even then, if you were on a different page it would save a preset even though you have just moved to a different page, that seems unintended?
> 
> This is why I prefer to create a separate issue for this, so we can discuss such things there 😄

I'll create one now

paescuj commented:
@br41nslug LGTY?

leon-marzahn committed:
Ignore page changes before saving presets

leon-marzahn commented:
@paescuj Here is the issue with PR attached https://github.com/directus/directus/issues/20202

leon-marzahn committed:
Fix prettier

br41nslug commented:
> This is why I prefer to create a separate issue for this, so we can discuss such things there 😄

Why would we split this up? In my current tests the current solution does not completely solve the issue of "Don't create new presets for users when they have made no changes" 🤔 

> I wouldn't consider this as a blocker for this PR though & prefer to track this in a separate issue.

I disagree, in my opinion this is exactly what the issue is about "creating new presets without changes" 🤔 Although i could be wrong, we're missing actual "reproduction steps" on the issue, perhaps we should add some more detail to the original image to prevent this confusion.

leon-marzahn commented:
> > This is why I prefer to create a separate issue for this, so we can discuss such things there 😄
> 
> Why would we split this up? In my current tests the current solution does not completely solve the issue of "Don't create new presets for users when they have made no changes" 🤔 Independently of wheter there are already presets defined or not.
> 
> > I wouldn't consider this as a blocker for this PR though & prefer to track this in a separate issue.
> 
> I disagree, in my opinion this is exactly what the issue is about "creating new presets without changes" 🤔 Although i could be wrong, we're missing actual "reproduction steps" on the issue, perhaps we should add some more detail to the original image to prevent this confusion.

If we rephrase the issue to exactly this and add reproduction steps, would that work for you? The only issue is that this PR (and the other one) might not cover every example and we would have to keep it open until all cases are fixed

br41nslug commented:
> might not cover every example and we would have to keep it open until all cases are fixed

We can never be sure to have covered all the possible cases but lets at least fix the 2 situations we know about.

leon-marzahn commented:
> > might not cover every example and we would have to keep it open until all cases are fixed
> 
> We can never be sure to have covered all the possible cases but lets at least fix the 2 situations we know about.

Okay, I will make changes to this issue and pr 👍🏻

leon-marzahn committed:
Merge branch 'no-preset-autosave-when-chaning-page' into no-preset-auto-save

leon-marzahn committed:
Remove changeset

br41nslug commented:
Or lets organize that in the second PR you've created as that already looks like it includes the changes from this PR and the extra fix for switching collections.

leon-marzahn commented:
> Or lets organize that in the second PR you've created as that already looks like it includes the changes from this PR and the extra fix for switching collections.

Oops, already deleted the other one 😅

leon-marzahn commented:
@br41nslug I've updated the issue to include a reproduction and the latest changes are in this branch, how does it look?

br41nslug commented:
Seems the situation i ran into is indeed resolved, will have a look at the code and test other layouts tomorrow

br41nslug commented on app/src/composables/use-preset.ts:
```diff
@@ -88,6 +88,14 @@ export function usePreset(
 
 	function updatePreset(preset: Partial<Preset>, immediate?: boolean) {
 		localPreset.value = assign({}, localPreset.value, preset);
+
+		const hasChanged = Object.keys(preset.layout_query ?? {}).some((key) => {
+			const query = (preset.layout_query ?? {})[key];
+			return !(Object.keys(query).length === 1 && 'page' in query);
+		});
+
+		if (!hasChanged) return;
+
 		immediate ? savePreset() : handleChanges();
 	}
```
```suggestion
	function checkIfChanged(existing: unknown, updated: unknown) {
		if (typeof existing !== 'object' || existing === null || typeof updated !== 'object' || updated === null) {
			return existing !== updated;
		}

		for (const key in updated) {
			if (key in existing === false) return true;
			const changed = checkIfChanged(existing[key], updated[key]);

			if (changed) {
				return true;
			}
		}

		return false;
	}

	function updatePreset(preset: Partial<Preset>, immediate?: boolean) {
		const hasChanged = checkIfChanged(localPreset.value, preset);
		if (!hasChanged) return;

		localPreset.value = assign({}, localPreset.value, preset);
		immediate ? savePreset() : handleChanges();
	}
```

leon-marzahn replied:
Won't this still save presets when the page changes? The new preset coming into `updatePreset` has a single key called `page` which i was trying to ignore

leon-marzahn replied:
@br41nslug Here is a recording of the fix i implemented, it appears to be working or is there a specific case that doesn't work?

[screen-capture.webm](https://github.com/directus/directus/assets/15183853/a4daa908-05de-4253-acf0-3fa7c8559597)

paescuj replied:
Not sure if this is the right place to handle this. If possible I'd solve this at the `layout` level, so it doesn't get passed to `updatePreset()` / isn't part of layout query in the first place. 

That being said, taking a step back, I think we need to decide whether we indeed want drop `page` from the presets. While that probably makes sense, I still feel like it's worth a separate issue, especially as it may have been used by some users so far. This is also reflected in the fact that we are already having a major discussion here.

The other two changes included in this PR (render order in calendar & check for change of `selectedSize`) are clearly bug fixes and could already be merged by now, IMO.

leon-marzahn replied:
> Not sure if this is the right place to handle this. If possible I'd solve this at the `layout` level, so it doesn't get passed to `updatePreset()` / isn't part of layout query in the first place.

That would make sense, as the page should be separate from layout in my opinion. 

> That being said, taking a step back, I think we need to decide whether we indeed want drop `page` from the presets. While that probably makes sense, I still feel like it's worth a separate issue, especially as it may have been used by some users so far. This is also reflected in the fact that we are already having a major discussion here.

We could open a discussion and ask more people, I reckon the biggest argument for that is the dynamic nature of pages and whether saving the current page actually makes sense. 

> The other two changes included in this PR (render order in calendar & check for change of `selectedSize`) are clearly bug fixes and could already be merged by now, IMO.

I agree

rijkvanzanten replied:
> That would make sense, as the page should be separate from layout in my opinion.

This definitely needs a bit more thinking. The original intent for persisting that information is persisting your place in the navigation when navigating around / logging in elsewhere. The current page of results is pretty important in that still, as you wouldn't want to be back on page 1 when you navigate in and out of a record 🤔

leon-marzahn replied:
> > That would make sense, as the page should be separate from layout in my opinion.
> 
> This definitely needs a bit more thinking. The original intent for persisting that information is persisting your place in the navigation when navigating around / logging in elsewhere. The current page of results is pretty important in that still, as you wouldn't want to be back on page 1 when you navigate in and out of a record 🤔

I wonder if we could just save an actual "local" preset in local storage? That would keep persistence in a session without saving a new preset which is very annoying in a multi user app

rijkvanzanten replied:
Worth thinking through! Lets discuss that in a feature request discussion though instead of a feature request that already partially implements something we aren't really sure about yet 😄

rijkvanzanten replied:
Yeah lets do it! The rest is looking good to me 🙂

azrikahar replied:
> so should I remove the change about the page? 

reverted via https://github.com/directus/directus/commit/bea5cf26122e9f9940588c0da21a1d64e08ee64d

br41nslug reviewed (dismissed):
The used fix for preset changes was a bit too effective 😬 It prevented saving layout changes and options altogether

rijkvanzanten committed:
Merge branch 'main' into no-preset-auto-save

rijkvanzanten committed:
Merge branch 'main' into no-preset-auto-save

azrikahar committed:
revert logic for detecting page changes

as per discussion in https://github.com/directus/directus/pull/20106/files#r1374908744

leon-marzahn commented:
Thanks @azrikahar! I was too busy with work 😅

azrikahar commented:
> Thanks @azrikahar! I was too busy with work 😅

No worries! The rest of the changes LGTM 😊
